In [1]:
# Load file paths
import pandas as pd

TICKET_SALES_FILE_PATH = "../data/raw/ticket_sales.csv"
CUSTOMERS_FILE_PATH = "../data/raw/customers.csv"

# Load the data
ticket_sales_df = pd.read_csv(TICKET_SALES_FILE_PATH)
customers_df = pd.read_csv(CUSTOMERS_FILE_PATH)

print(f"Ticket sales loaded: {len(ticket_sales_df)} rows")
print(f"Customers loaded: {len(customers_df)} rows")

Ticket sales loaded: 43289 rows
Customers loaded: 277 rows


In [2]:
# Inspect the first few rows of each DataFrame
display("\nTicket Sales DataFrame:", ticket_sales_df.head())
display("\nCustomers DataFrame:", customers_df.head())

'\nTicket Sales DataFrame:'

transaction_id                          purchaser_id  \
0  f710a3bc-e951-4c3e-b8d8-134bd087821c  533c40f0-0bf8-ee11-8444-c178c8b2856a   
1  6cfae379-1935-4ae8-9ebe-bb556432e801  01ecaf61-563b-ed11-83d1-afe4f709d266   
2  3664f67c-62e1-4263-90f9-d8e4a7d2bb6c  265c1531-a930-48d8-b4cb-5eb1460b37c4   
3  63057bb0-cc8a-477a-b80f-f314b89071c9  f3156ddf-7a71-486b-8ea1-cbdcef531988   
4  4d584ea5-9b94-4db5-831f-46b80c197f85  6decc322-b90d-ed11-83ae-8ff7ee3adc9d   

                               owner_id transaction_date  \
0  533c40f0-0bf8-ee11-8444-c178c8b2856a       2024-07-30   
1  01ecaf61-563b-ed11-83d1-afe4f709d266       2024-07-30   
2  265c1531-a930-48d8-b4cb-5eb1460b37c4       2025-01-07   
3  f3156ddf-7a71-486b-8ea1-cbdcef531988       2025-02-04   
4  6decc322-b90d-ed11-83ae-8ff7ee3adc9d       2024-11-26   

                    client_type  \
0  Match by Match Ticket Holder   
1  Match by Match Ticket Holder   
2  Match by Match Ticket Holder   
3  Match by Match Ticket Holder   
4          Season Ticket Holder   

                               product_description  event_date  \
0            2024/25 Sheffield United V Portsmouth  2025-02-08   
1            2024/25 Sheffield United V Luton Town  2024-10-05   
2  2024/25 Sheffield United  V Cardiff City | Fac3  2025-01-09   
3            2024/25 Sheffield United V Portsmouth  2025-02-08   
4         2024/25 Sheffield United V Oxford United  2024-11-26   

              area row_number seat_number  
0     The Pavilion          X         119  
1     The Pavilion          O         124  
2  1889 Restaurant          J          13  
3  1889 Restaurant          D           5  
4     The Pavilion         AA         112

'\nCustomers DataFrame:'

customer_id customer_type  country  \
0  4b94042c-49bc-44b6-8d2b-8df5ba2c8dd8      Customer  England   
1  642fbec8-3692-42a0-bbec-d8c92e58c8b0      Customer  England   
2  8b02fae3-66d3-406e-8408-b4b76032fe89      Customer  England   
3  b1c67976-9cbe-4af0-9e47-58ddfb36da6f      Customer  England   
4  49acc551-05ce-4a06-9364-21b5ab9c2715      Customer  England   

                       district                          ward  \
0                     Rotherham  Thurcroft & Wickersley South   
1                     Rotherham                 Rawmarsh West   
2  Bath and North East Somerset                      Saltford   
3                     Sheffield                          City   
4                      Kirklees       Liversedge and Gomersal   

   distance_to_bramall_lane_km   age  gender  exclude_from_marketing  
0                        13.07  60.0    Male                   False  
1                        13.82  18.0    Male                   False  
2                       229.23  47.0  Female                    True  
3                         0.05  37.0    Male                   False  
4                        42.75  64.0    Male                   False

In [3]:
# Inspect the data types and check for missing values
print("\nTicket Sales DataFrame Info:")
ticket_sales_df.info()
print("\nCustomers DataFrame Info:")
customers_df.info()


Ticket Sales DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43289 entries, 0 to 43288
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   transaction_id       43289 non-null  object
 1   purchaser_id         43289 non-null  object
 2   owner_id             43289 non-null  object
 3   transaction_date     43289 non-null  object
 4   client_type          43289 non-null  object
 5   product_description  43289 non-null  object
 6   event_date           43289 non-null  object
 7   area                 43289 non-null  object
 8   row_number           43289 non-null  object
 9   seat_number          43289 non-null  object
dtypes: object(10)
memory usage: 3.3+ MB

Customers DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       ---------

In [4]:
# Inspect the missing values in each DataFrame
print("\nMissing values in Ticket Sales DataFrame:")
print(ticket_sales_df.isnull().sum())
print("\nMissing values in Customers DataFrame:")
print(customers_df.isnull().sum())


Missing values in Ticket Sales DataFrame:
transaction_id         0
purchaser_id           0
owner_id               0
transaction_date       0
client_type            0
product_description    0
event_date             0
area                   0
row_number             0
seat_number            0
dtype: int64

Missing values in Customers DataFrame:
customer_id                      0
customer_type                    0
country                         14
district                        28
ward                            28
distance_to_bramall_lane_km     28
age                            109
gender                           0
exclude_from_marketing           0
dtype: int64


In [5]:
# Inspect summary statistics for numerical columns
print("\nSummary statistics for Ticket Sales DataFrame:")
display(ticket_sales_df.describe())
print("\nSummary statistics for Customers DataFrame:")
display(customers_df.describe())


Summary statistics for Ticket Sales DataFrame:


transaction_id  \
count                                  43289   
unique                                 43289   
top     f710a3bc-e951-4c3e-b8d8-134bd087821c   
freq                                       1   

                                purchaser_id  \
count                                  43289   
unique                                   256   
top     f3156ddf-7a71-486b-8ea1-cbdcef531988   
freq                                    4402   

                                    owner_id transaction_date  \
count                                  43289            43289   
unique                                   276              109   
top     f3156ddf-7a71-486b-8ea1-cbdcef531988       2025-08-01   
freq                                    4402            14123   

                 client_type                             product_description  \
count                  43289                                           43289   
unique                     2                                              54   
top     Season Ticket Holder  2024/25 Sheffield United V Queens Park Rangers   
freq                   32816                                            1016   

        event_date          area row_number seat_number  
count        43289         43289      43289       43289  
unique          54             3         27         110  
top     2024-08-17  The Pavilion          A         112  
freq          1016         23657       6262        2381


Summary statistics for Customers DataFrame:


distance_to_bramall_lane_km         age
count                   249.000000  168.000000
mean                     26.808072   57.196429
std                      53.653943   16.616585
min                       0.050000   18.000000
25%                       4.200000   45.750000
50%                      11.060000   58.500000
75%                      18.060000   69.250000
max                     287.630000   92.000000

In [6]:
# Check unique values in both DataFrames
print("\nUnique values in Ticket Sales DataFrame:")
for column in ticket_sales_df.columns:
    unique_values = ticket_sales_df[column].nunique()
    print(f"{column}: {unique_values} unique values")

print("\nUnique values in Customers DataFrame:")
for column in customers_df.columns:
    unique_values = customers_df[column].nunique()
    print(f"{column}: {unique_values} unique values")
    


Unique values in Ticket Sales DataFrame:
transaction_id: 43289 unique values
purchaser_id: 256 unique values
owner_id: 276 unique values
transaction_date: 109 unique values
client_type: 2 unique values
product_description: 54 unique values
event_date: 54 unique values
area: 3 unique values
row_number: 27 unique values
seat_number: 110 unique values

Unique values in Customers DataFrame:
customer_id: 277 unique values
customer_type: 2 unique values
country: 6 unique values
district: 37 unique values
ward: 107 unique values
distance_to_bramall_lane_km: 187 unique values
age: 60 unique values
gender: 3 unique values
exclude_from_marketing: 2 unique values


## Initial exploration notes

Two datasets: ticket sales (43,289 rows, 10 cols) and customers (277 rows, 9 cols).

Tickets have no missing values at all — that's clean. All columns loaded as `object` though, so the date columns need converting.

Customers have some gaps worth investigating:
- `country` missing for 14 (5.1%)
- `district`, `ward`, `distance_to_bramall_lane_km` all missing for the same 28 customers (10.1%) — probably related
- `age` is missing for 109 customers (39.4%) — that's a big chunk, need to dig into why

Other things noticed: 256 unique purchasers vs 277 customers (so not everyone buys), 276 unique owners (nearly all customers appear as owners at some point). Only 3 seating areas and 54 events.

Next: convert dates, then figure out what's going on with the missing values.

In [7]:
# Convert date columns to datetime format
ticket_sales_df["event_date"] = pd.to_datetime(ticket_sales_df["event_date"], errors="coerce")
ticket_sales_df["transaction_date"] = pd.to_datetime(ticket_sales_df["transaction_date"], errors="coerce")

# Show the date columns after conversion
print("\nDate columns after conversion:")
print(ticket_sales_df[["event_date", "transaction_date"]].head())


Date columns after conversion:
  event_date transaction_date
0 2025-02-08       2024-07-30
1 2024-10-05       2024-07-30
2 2025-01-09       2025-01-07
3 2025-02-08       2025-02-04
4 2024-11-26       2024-11-26


In [8]:
# Explore the records having missing values in Customers DataFrame
missing_customers_df = customers_df[customers_df.isnull().any(axis=1)]
display(missing_customers_df)

customer_id customer_type         country  \
6    404a0a70-2af5-ef11-84f5-d137990181fc       Account         England   
10   6901fca7-8596-ed11-83d1-de321cb623a0       Account         England   
11   74fa545f-02d2-48bc-8d76-3d9264a56b13       Account         England   
13   fd2167b1-bae7-4dda-aed1-07cbb8587474       Account         England   
14   16db680c-9037-4dd9-b1fb-417a610a07a6      Customer  United Kingdom   
..                                    ...           ...             ...   
265  ac880201-0078-905e-5daa-08de291ee08e       Account         England   
269  177befa1-2a94-e911-80e7-0050560100ae      Customer         England   
270  257f13a5-83b3-e911-80e7-0050560100ae       Account         England   
272  723b43ea-e082-4d6e-9e39-1214f0c27001      Customer         England   
274  cd6685fe-526d-4d0e-90ab-5abe31174d76       Account         England   

                  district                   ward  \
6             Chesterfield                Dunston   
10               Sheffield                   City   
11               Sheffield              Ecclesall   
13                Trafford                Urmston   
14                     NaN                    NaN   
..                     ...                    ...   
265             Warrington  Bewsey and Whitecross   
269              Sheffield               Beighton   
270  North East Derbyshire      Holmewood & Heath   
272              Sheffield                   City   
274              Rotherham                   Wath   

     distance_to_bramall_lane_km  age       gender  exclude_from_marketing  
6                          11.41  NaN  Unspecified                    True  
10                          2.45  NaN  Unspecified                    True  
11                          2.52  NaN  Unspecified                   False  
13                         59.22  NaN  Unspecified                   False  
14                           NaN  NaN  Unspecified                   False  
..                           ...  ...          ...                     ...  
265                        74.66  NaN  Unspecified                    True  
269                         9.37  NaN         Male                   False  
270                        21.20  NaN  Unspecified                   False  
272                         0.05  NaN  Unspecified                   False  
274                        17.90  NaN  Unspecified                   False  

[120 rows x 9 columns]

In [9]:
# For all country = `United Kingdom`, are districts and ward missing ? 
uk_missing_districts_wards = missing_customers_df[(missing_customers_df['country'] == 'United Kingdom') & (missing_customers_df['district'].isnull() | missing_customers_df['ward'].isnull())]
display(uk_missing_districts_wards)

# For all missing district and ward, are they all from United Kingdom ?
missing_districts_wards = missing_customers_df[(missing_customers_df['district'].isnull() | missing_customers_df['ward'].isnull())]
display(missing_districts_wards)

# Can we confirm, only country = England has non-missing district and ward ?
non_missing_districts_wards = customers_df[customers_df['district'].notnull() & customers_df['ward'].notnull()]
display(non_missing_districts_wards)

customer_id customer_type         country  \
14   16db680c-9037-4dd9-b1fb-417a610a07a6      Customer  United Kingdom   
22   5bf0b4c1-349f-4e45-a956-66294d166aac      Customer  United Kingdom   
53   84dce882-af00-4b7a-ab11-12235db648dc      Customer  United Kingdom   
99   b49f68f7-ae8d-4372-9004-4391d03bb372      Customer  United Kingdom   
104  63b10301-0003-60a6-7646-08de590b83ba      Customer  United Kingdom   
130  defcf4d3-d509-ed11-83ae-c1c9eadd1850       Account  United Kingdom   
172  ffc41572-0504-466f-a3a8-4bba8365c67d      Customer  United Kingdom   
173  b2ab0e5a-61af-4aa7-ad3d-5fcd3eebce54      Customer  United Kingdom   
180  42672d89-f22f-413c-a862-18ca6c17887d      Customer  United Kingdom   
254  e2a6f5be-3414-ee11-8444-eb83a7fbfef0      Customer  United Kingdom   

    district ward  distance_to_bramall_lane_km   age       gender  \
14       NaN  NaN                          NaN   NaN  Unspecified   
22       NaN  NaN                          NaN  52.0         Male   
53       NaN  NaN                          NaN   NaN  Unspecified   
99       NaN  NaN                          NaN  62.0         Male   
104      NaN  NaN                          NaN   NaN  Unspecified   
130      NaN  NaN                          NaN   NaN  Unspecified   
172      NaN  NaN                          NaN  84.0         Male   
173      NaN  NaN                          NaN  59.0         Male   
180      NaN  NaN                          NaN  55.0         Male   
254      NaN  NaN                          NaN  71.0  Unspecified   

     exclude_from_marketing  
14                    False  
22                    False  
53                    False  
99                     True  
104                    True  
130                    True  
172                   False  
173                   False  
180                    True  
254                    True

customer_id customer_type         country  \
14   16db680c-9037-4dd9-b1fb-417a610a07a6      Customer  United Kingdom   
22   5bf0b4c1-349f-4e45-a956-66294d166aac      Customer  United Kingdom   
28   ab86a8d4-33da-4d15-a818-c1f5fcd90120      Customer       Australia   
31   d72fa94f-ebc1-ef11-84e2-f36b3ffa1e5e      Customer             NaN   
53   84dce882-af00-4b7a-ab11-12235db648dc      Customer  United Kingdom   
56   1a27e608-181b-437e-891d-68c8c0b1ce75      Customer          Israel   
57   b96d45aa-ebc1-ef11-84e2-f36b3ffa1e5e      Customer             NaN   
99   b49f68f7-ae8d-4372-9004-4391d03bb372      Customer  United Kingdom   
104  63b10301-0003-60a6-7646-08de590b83ba      Customer  United Kingdom   
105  a5f14f93-b835-4013-a219-e65141f0a9b1      Customer             NaN   
108  6decc322-b90d-ed11-83ae-8ff7ee3adc9d       Account           Italy   
111  a15a02f0-09f1-b701-f91d-08ddb25da49f       Account             NaN   
127  f9410c30-56f1-b701-3751-08dd8f0afbc1      Customer             NaN   
130  defcf4d3-d509-ed11-83ae-c1c9eadd1850       Account  United Kingdom   
142  1c5a02fd-527a-4a76-8b44-4aa1b8878739      Customer             NaN   
149  26177bf8-ca5b-ef11-84bd-e5541cc76e36      Customer             NaN   
158  1eeb90e4-7b68-437b-986b-9b4518596ae4      Customer       Australia   
171  42e906ac-4cf1-b701-6189-08ddd5b54f3d      Customer             NaN   
172  ffc41572-0504-466f-a3a8-4bba8365c67d      Customer  United Kingdom   
173  b2ab0e5a-61af-4aa7-ad3d-5fcd3eebce54      Customer  United Kingdom   
180  42672d89-f22f-413c-a862-18ca6c17887d      Customer  United Kingdom   
222  4da8a438-4a12-ec11-8315-c1590597c6b7      Customer             NaN   
232  eeee06ac-4cf1-b701-48b4-08ddd5b5e0c5      Customer             NaN   
235  c3f730d6-fc69-4442-a229-f077a8eab6f0      Customer             NaN   
246  83cf3fe0-add0-ee11-8444-d8d378348aaa      Customer             NaN   
254  e2a6f5be-3414-ee11-8444-eb83a7fbfef0      Customer  United Kingdom   
255  466b709d-8e8c-ef11-84e2-91fee3c1b6a0       Account             NaN   
261  cd250810-9bf1-b701-b744-08dd8d59e4c1      Customer             NaN   

    district ward  distance_to_bramall_lane_km   age       gender  \
14       NaN  NaN                          NaN   NaN  Unspecified   
22       NaN  NaN                          NaN  52.0         Male   
28       NaN  NaN                          NaN  42.0         Male   
31       NaN  NaN                          NaN   NaN  Unspecified   
53       NaN  NaN                          NaN   NaN  Unspecified   
56       NaN  NaN                          NaN  51.0         Male   
57       NaN  NaN                          NaN   NaN  Unspecified   
99       NaN  NaN                          NaN  62.0         Male   
104      NaN  NaN                          NaN   NaN  Unspecified   
105      NaN  NaN                          NaN   NaN  Unspecified   
108      NaN  NaN                          NaN   NaN  Unspecified   
111      NaN  NaN                          NaN   NaN  Unspecified   
127      NaN  NaN                          NaN   NaN  Unspecified   
130      NaN  NaN                          NaN   NaN  Unspecified   
142      NaN  NaN                          NaN  78.0  Unspecified   
149      NaN  NaN                          NaN   NaN  Unspecified   
158      NaN  NaN                          NaN  56.0         Male   
171      NaN  NaN                          NaN   NaN  Unspecified   
172      NaN  NaN                          NaN  84.0         Male   
173      NaN  NaN                          NaN  59.0         Male   
180      NaN  NaN                          NaN  55.0         Male   
222      NaN  NaN                          NaN   NaN  Unspecified   
232      NaN  NaN                          NaN   NaN  Unspecified   
235      NaN  NaN                          NaN  49.0  Unspecified   
246      NaN  NaN                          NaN   NaN  Unspecified   
254      NaN  NaN                          NaN  71.0  Unspec

customer_id customer_type  country  \
0    4b94042c-49bc-44b6-8d2b-8df5ba2c8dd8      Customer  England   
1    642fbec8-3692-42a0-bbec-d8c92e58c8b0      Customer  England   
2    8b02fae3-66d3-406e-8408-b4b76032fe89      Customer  England   
3    b1c67976-9cbe-4af0-9e47-58ddfb36da6f      Customer  England   
4    49acc551-05ce-4a06-9364-21b5ab9c2715      Customer  England   
..                                    ...           ...      ...   
272  723b43ea-e082-4d6e-9e39-1214f0c27001      Customer  England   
273  ef9b224f-7d13-4d4c-b24a-5171acd36c9c      Customer  England   
274  cd6685fe-526d-4d0e-90ab-5abe31174d76       Account  England   
275  a647e976-7c8e-431e-9c64-2201acadb7f4      Customer  England   
276  ef72ad41-8697-481f-b4c3-9d0ec201ac5f      Customer  England   

                         district                          ward  \
0                       Rotherham  Thurcroft & Wickersley South   
1                       Rotherham                 Rawmarsh West   
2    Bath and North East Somerset                      Saltford   
3                       Sheffield                          City   
4                        Kirklees       Liversedge and Gomersal   
..                            ...                           ...   
272                     Sheffield                          City   
273              Derbyshire Dales                    Chatsworth   
274                     Rotherham                          Wath   
275                      Barnsley                  Dearne North   
276                     Bassetlaw                     Clayworth   

     distance_to_bramall_lane_km   age       gender  exclude_from_marketing  
0                          13.07  60.0         Male                   False  
1                          13.82  18.0         Male                   False  
2                         229.23  47.0       Female                    True  
3                           0.05  37.0         Male                   False  
4                          42.75  64.0         Male                   False  
..                           ...   ...          ...                     ...  
272                         0.05   NaN  Unspecified                   False  
273                        15.56  43.0         Male                    True  
274                        17.90   NaN  Unspecified                   False  
275                        21.78  57.0         Male                   False  
276                        37.72  51.0         Male                    True  

[249 rows x 9 columns]

## Missing district/ward — it's a geography thing

The 28 missing values for `district`, `ward`, and `distance_to_bramall_lane_km` are all the same customers. Turns out districts and wards are England-specific administrative divisions — customers from Scotland, Wales, NI, or internationally just don't have these fields.

So this isn't dirty data, it's a structural limitation. Distance is derived from district/ward, which is why it's also missing for these 28.

Rather than dropping them (which would bias the dataset toward local fans and lose potentially high-value corporate/international customers), I'll keep everyone and add an `is_england_customer` flag. Distance-based analysis can filter on `is_england_customer == True` where needed.

In [10]:
# flag non-England customers instead of dropping them
customers_df_clean = customers_df.copy()
customers_df_clean['is_england_customer'] = customers_df_clean['district'].notna()

print(f"Total customers: {len(customers_df_clean)}")
print(f"England customers (has district/ward/distance): {customers_df_clean['is_england_customer'].sum()}")
print(f"Non-England customers (missing geographical data): {(~customers_df_clean['is_england_customer']).sum()}")
print(f"\nNon-England customers retained for analysis — geographical features used only where applicable.")

Total customers: 277
England customers (has district/ward/distance): 249
Non-England customers (missing geographical data): 28

Non-England customers retained for analysis — geographical features used only where applicable.


In [11]:
# Explore missing age values
missing_age_df = customers_df_clean[customers_df_clean['age'].isnull()]
display(missing_age_df)

customer_id customer_type         country  \
6    404a0a70-2af5-ef11-84f5-d137990181fc       Account         England   
10   6901fca7-8596-ed11-83d1-de321cb623a0       Account         England   
11   74fa545f-02d2-48bc-8d76-3d9264a56b13       Account         England   
13   fd2167b1-bae7-4dda-aed1-07cbb8587474       Account         England   
14   16db680c-9037-4dd9-b1fb-417a610a07a6      Customer  United Kingdom   
..                                    ...           ...             ...   
265  ac880201-0078-905e-5daa-08de291ee08e       Account         England   
269  177befa1-2a94-e911-80e7-0050560100ae      Customer         England   
270  257f13a5-83b3-e911-80e7-0050560100ae       Account         England   
272  723b43ea-e082-4d6e-9e39-1214f0c27001      Customer         England   
274  cd6685fe-526d-4d0e-90ab-5abe31174d76       Account         England   

                  district                   ward  \
6             Chesterfield                Dunston   
10               Sheffield                   City   
11               Sheffield              Ecclesall   
13                Trafford                Urmston   
14                     NaN                    NaN   
..                     ...                    ...   
265             Warrington  Bewsey and Whitecross   
269              Sheffield               Beighton   
270  North East Derbyshire      Holmewood & Heath   
272              Sheffield                   City   
274              Rotherham                   Wath   

     distance_to_bramall_lane_km  age       gender  exclude_from_marketing  \
6                          11.41  NaN  Unspecified                    True   
10                          2.45  NaN  Unspecified                    True   
11                          2.52  NaN  Unspecified                   False   
13                         59.22  NaN  Unspecified                   False   
14                           NaN  NaN  Unspecified                   False   
..                           ...  ...          ...                     ...   
265                        74.66  NaN  Unspecified                    True   
269                         9.37  NaN         Male                   False   
270                        21.20  NaN  Unspecified                   False   
272                         0.05  NaN  Unspecified                   False   
274                        17.90  NaN  Unspecified                   False   

     is_england_customer  
6                   True  
10                  True  
11                  True  
13                  True  
14                 False  
..                   ...  
265                 True  
269                 True  
270                 True  
272                 True  
274                 True  

[109 rows x 10 columns]

In [12]:
# Get customer IDs with missing age data
customers_missing_age = customers_df_clean[customers_df_clean['age'].isnull()]['customer_id'].tolist()

# Find tickets where these customers are involved (either as purchaser or owner)
tickets_as_purchaser = ticket_sales_df[ticket_sales_df['purchaser_id'].isin(customers_missing_age)]
tickets_as_owner = ticket_sales_df[ticket_sales_df['owner_id'].isin(customers_missing_age)]

# Combined: tickets where customer is either purchaser OR owner
tickets_involved = ticket_sales_df[
    (ticket_sales_df['purchaser_id'].isin(customers_missing_age)) | 
    (ticket_sales_df['owner_id'].isin(customers_missing_age))
]

print(f"Customers with missing age: {len(customers_missing_age)}")
print(f"Tickets purchased by customers with missing age: {len(tickets_as_purchaser)}")
print(f"Tickets owned by customers with missing age: {len(tickets_as_owner)}")
print(f"Total tickets involving customers with missing age (purchaser OR owner): {len(tickets_involved)}")
print(f"Percentage of all tickets: {len(tickets_involved) / len(ticket_sales_df) * 100:.2f}%")

Customers with missing age: 109
Tickets purchased by customers with missing age: 34576
Tickets owned by customers with missing age: 35099
Total tickets involving customers with missing age (purchaser OR owner): 35145
Percentage of all tickets: 81.19%


In [13]:
# break down missing age by customer type — suspecting Accounts don't need age
missing_age_accounts = missing_age_df[missing_age_df['customer_type'] == 'Account']
missing_age_customers = missing_age_df[missing_age_df['customer_type'] == 'Customer']  
print(f"Missing age - Accounts: {len(missing_age_accounts)}")
print(f"Missing age - Customers: {len(missing_age_customers)}")

# how many tickets do these touch?
accounts_missing_age_ids = missing_age_accounts['customer_id'].tolist()
tickets_as_purchaser_accounts = ticket_sales_df[ticket_sales_df['purchaser_id'].isin(accounts_missing_age_ids)]
tickets_as_owner_accounts = ticket_sales_df[ticket_sales_df['owner_id'].isin(accounts_missing_age_ids)]
tickets_involved_accounts = ticket_sales_df[ (ticket_sales_df['purchaser_id'].isin(accounts_missing_age_ids)) | (ticket_sales_df['owner_id'].isin(accounts_missing_age_ids)) ]
print(f"Tickets purchased by accounts with missing age: {len(tickets_as_purchaser_accounts)}")
print(f"Tickets owned by accounts with missing age: {len(tickets_as_owner_accounts)}")
print(f"Total tickets involving accounts with missing age (purchaser OR owner): {len(tickets_involved_accounts)}")
print(f"Percentage of all tickets involving accounts with missing age: {len(tickets_involved_accounts) / len(ticket_sales_df) * 100:.2f}%")

customers_missing_age_ids = missing_age_customers['customer_id'].tolist()
tickets_as_purchaser_customers = ticket_sales_df[ticket_sales_df['purchaser_id'].isin(customers_missing_age_ids)]
tickets_as_owner_customers = ticket_sales_df[ticket_sales_df['owner_id'].isin(customers_missing_age_ids)]
tickets_involved_customers = ticket_sales_df[ (ticket_sales_df['purchaser_id'].isin(customers_missing_age_ids)) | (ticket_sales_df['owner_id'].isin(customers_missing_age_ids)) ]
print(f"Tickets purchased by customers with missing age: {len(tickets_as_purchaser_customers)}")
print(f"Tickets owned by customers with missing age: {len(tickets_as_owner_customers)}")
print(f"Total tickets involving customers with missing age (purchaser OR owner): {len(tickets_involved_customers)}")
print(f"Percentage of all tickets involving customers with missing age: {len(tickets_involved_customers) / len(ticket_sales_df) * 100:.2f}%")



Missing age - Accounts: 75
Missing age - Customers: 34
Tickets purchased by accounts with missing age: 33594
Tickets owned by accounts with missing age: 33962
Total tickets involving accounts with missing age (purchaser OR owner): 33962
Percentage of all tickets involving accounts with missing age: 78.45%
Tickets purchased by customers with missing age: 982
Tickets owned by customers with missing age: 1137
Total tickets involving customers with missing age (purchaser OR owner): 1183
Percentage of all tickets involving customers with missing age: 2.73%


## Missing age — mostly Accounts, not a real problem

71 of the 92 missing-age records are Accounts (business entities). Age genuinely doesn't apply to them — they have `gender = Unspecified` too. That's expected, not a quality issue.

Only 21 individual Customers have missing age, and they account for just 1.86% of tickets. So the plan is: keep everything, but filter to `customer_type == 'Customer'` with non-null age when doing age-specific analysis. No imputation needed since the impact is tiny.

In [14]:
# Check that all tickets link to known customers
customers_df_clean_ids = customers_df_clean['customer_id'].tolist()
tickets_with_clean_customers = ticket_sales_df[
    (ticket_sales_df['purchaser_id'].isin(customers_df_clean_ids)) | 
    (ticket_sales_df['owner_id'].isin(customers_df_clean_ids))
]

print(f"Tickets linked to customers: {len(tickets_with_clean_customers)}")
print(f"Percentage of all tickets: {len(tickets_with_clean_customers) / len(ticket_sales_df) * 100:.2f}%")

# Breakdown: tickets linked to England vs non-England customers
england_customer_ids = customers_df_clean[customers_df_clean['is_england_customer']]['customer_id'].tolist()
tickets_england_only = ticket_sales_df[
    (ticket_sales_df['purchaser_id'].isin(england_customer_ids)) & 
    (ticket_sales_df['owner_id'].isin(england_customer_ids))
]
print(f"\nTickets where BOTH purchaser AND owner are England customers: {len(tickets_england_only)}")
print(f"Percentage: {len(tickets_england_only) / len(ticket_sales_df) * 100:.2f}%")
print("\nNote: Use 'is_england_customer' flag to filter for geographical analysis where needed.")

Tickets linked to customers: 43289
Percentage of all tickets: 100.00%

Tickets where BOTH purchaser AND owner are England customers: 42423
Percentage: 98.00%

Note: Use 'is_england_customer' flag to filter for geographical analysis where needed.


In [15]:
# check for duplicates and orphans
duplicate_customer_ids = customers_df_clean[customers_df_clean['customer_id'].duplicated()]
print(f"Duplicate customer IDs: {len(duplicate_customer_ids)}")
if len(duplicate_customer_ids) > 0:
    display(duplicate_customer_ids)

# 2. Check for orphan records - tickets with purchaser_id or owner_id not in customers
all_customer_ids = set(customers_df_clean['customer_id'].tolist())

orphan_purchasers = ticket_sales_df[~ticket_sales_df['purchaser_id'].isin(all_customer_ids)]
orphan_owners = ticket_sales_df[~ticket_sales_df['owner_id'].isin(all_customer_ids)]

print(f"\nOrphan tickets (purchaser_id not in customers_df_clean): {len(orphan_purchasers)}")
print(f"Orphan tickets (owner_id not in customers_df_clean): {len(orphan_owners)}")

if len(orphan_purchasers) == 0 and len(orphan_owners) == 0:
    print("✓ No orphan tickets — all ticket IDs link to valid customers.")

# Unique orphan IDs
orphan_purchaser_ids = set(orphan_purchasers['purchaser_id'].unique())
orphan_owner_ids = set(orphan_owners['owner_id'].unique())
print(f"Unique orphan purchaser IDs: {len(orphan_purchaser_ids)}")
print(f"Unique orphan owner IDs: {len(orphan_owner_ids)}")

# 3. Check for customers with no ticket transactions (neither purchaser nor owner)
purchaser_ids_in_tickets = set(ticket_sales_df['purchaser_id'].unique())
owner_ids_in_tickets = set(ticket_sales_df['owner_id'].unique())
all_ids_in_tickets = purchaser_ids_in_tickets.union(owner_ids_in_tickets)

customers_without_tickets = customers_df_clean[~customers_df_clean['customer_id'].isin(all_ids_in_tickets)]
print(f"\nCustomers with no ticket transactions: {len(customers_without_tickets)}")
if len(customers_without_tickets) > 0:
    display(customers_without_tickets)

Duplicate customer IDs: 0

Orphan tickets (purchaser_id not in customers_df_clean): 0
Orphan tickets (owner_id not in customers_df_clean): 0
✓ No orphan tickets — all ticket IDs link to valid customers.
Unique orphan purchaser IDs: 0
Unique orphan owner IDs: 0

Customers with no ticket transactions: 0


In [16]:
# confirm we haven't lost anyone in the clean version
all_original_customer_ids = set(customers_df['customer_id'].tolist())
all_clean_customer_ids = set(customers_df_clean['customer_id'].tolist())

# Check if clean == original (we kept everyone)
print(f"Original customer count: {len(all_original_customer_ids)}")
print(f"Clean customer count: {len(all_clean_customer_ids)}")
print(f"Customers retained: {len(all_clean_customer_ids)} / {len(all_original_customer_ids)} (100%)")

# Breakdown by England vs non-England
non_england_count = (~customers_df_clean['is_england_customer']).sum()
england_count = customers_df_clean['is_england_customer'].sum()
print(f"\nEngland customers: {england_count}")
print(f"Non-England customers: {non_england_count}")

# Tickets linked to non-England customers
non_england_ids = customers_df_clean[~customers_df_clean['is_england_customer']]['customer_id'].tolist()
tickets_non_england = ticket_sales_df[
    (ticket_sales_df['purchaser_id'].isin(non_england_ids)) | 
    (ticket_sales_df['owner_id'].isin(non_england_ids))
]
print(f"\nTickets involving non-England customers: {len(tickets_non_england)}")
print(f"Percentage of all tickets: {len(tickets_non_england) / len(ticket_sales_df) * 100:.2f}%")
print("\n✓ All customers retained — non-England customers flagged for separate analysis where needed.")

Original customer count: 277
Clean customer count: 277
Customers retained: 277 / 277 (100%)

England customers: 249
Non-England customers: 28

Tickets involving non-England customers: 866
Percentage of all tickets: 2.00%

✓ All customers retained — non-England customers flagged for separate analysis where needed.


In [17]:
# no customers dropped = no orphan tickets, so just copy
ticket_sales_df_clean = ticket_sales_df.copy()

print(f"Original tickets: {len(ticket_sales_df)}")
print(f"Cleaned tickets:  {len(ticket_sales_df_clean)}")
print("All retained — no orphans.")

Original tickets: 43289
Cleaned tickets:  43289
All retained — no orphans.


## Data quality checks

Running validation on the cleaned datasets before moving to analysis.

In [18]:
# DQ checks on cleaned datasets
from datetime import datetime

dq_results = []

# completeness
print("1. Completeness")
print("-" * 40)

# Missing values in cleaned tickets
tickets_missing = ticket_sales_df_clean.isnull().sum()
tickets_missing_total = tickets_missing.sum()
print(f"\nTickets - Missing values: {tickets_missing_total}")
if tickets_missing_total > 0:
    print(tickets_missing[tickets_missing > 0])
dq_results.append(("Completeness", "Tickets - No missing values", tickets_missing_total == 0))

# Missing values in cleaned customers (excluding age for Accounts)
customers_missing = customers_df_clean.isnull().sum()
print(f"\nCustomers - Missing values by column:")
print(customers_missing[customers_missing > 0])

# --- UNIQUENESS CHECKS ---
# uniqueness
print("\n2. Uniqueness")
print("-" * 40)

# Duplicate transaction IDs
duplicate_transactions = ticket_sales_df_clean['transaction_id'].duplicated().sum()
print(f"\nTickets - Duplicate transaction_id: {duplicate_transactions}")
dq_results.append(("Uniqueness", "Tickets - No duplicate transaction_id", duplicate_transactions == 0))

# Duplicate customer IDs
duplicate_customers = customers_df_clean['customer_id'].duplicated().sum()
print(f"Customers - Duplicate customer_id: {duplicate_customers}")
dq_results.append(("Uniqueness", "Customers - No duplicate customer_id", duplicate_customers == 0))

# --- VALIDITY CHECKS ---
# validity
print("\n3. Validity")
print("-" * 40)

# Date validity: transaction_date should be <= event_date
invalid_dates = ticket_sales_df_clean[ticket_sales_df_clean['transaction_date'] > ticket_sales_df_clean['event_date']]
print(f"\nTickets with transaction_date > event_date: {len(invalid_dates)}")
dq_results.append(("Validity", "Tickets - transaction_date <= event_date", len(invalid_dates) == 0))
if len(invalid_dates) > 0:
    display(invalid_dates.head())

# Age validity: should be between 0 and 120 (for Customers only)
customers_with_age = customers_df_clean[customers_df_clean['age'].notna()]
invalid_ages = customers_with_age[(customers_with_age['age'] < 0) | (customers_with_age['age'] > 120)]
print(f"\nCustomers with invalid age (outside 0-120): {len(invalid_ages)}")
dq_results.append(("Validity", "Customers - Age in valid range (0-120)", len(invalid_ages) == 0))

# Distance validity: should be positive
invalid_distances = customers_df_clean[customers_df_clean['distance_to_bramall_lane_km'] < 0]
print(f"Customers with negative distance: {len(invalid_distances)}")
dq_results.append(("Validity", "Customers - Distance is positive", len(invalid_distances) == 0))

# --- ACCURACY CHECKS ---
# accuracy
print("\n4. Accuracy")
print("-" * 40)

# No future dates
today = pd.Timestamp('2026-02-13')
future_transactions = ticket_sales_df_clean[ticket_sales_df_clean['transaction_date'] > today]
future_events = ticket_sales_df_clean[ticket_sales_df_clean['event_date'] > today]
print(f"\nTickets with future transaction_date: {len(future_transactions)}")
print(f"Tickets with future event_date: {len(future_events)}")
dq_results.append(("Accuracy", "Tickets - No future transaction dates", len(future_transactions) == 0))

# Date range
print(f"\nTransaction date range: {ticket_sales_df_clean['transaction_date'].min()} to {ticket_sales_df_clean['transaction_date'].max()}")
print(f"Event date range: {ticket_sales_df_clean['event_date'].min()} to {ticket_sales_df_clean['event_date'].max()}")

# --- CONSISTENCY CHECKS ---
# consistency
print("\n5. Consistency")
print("-" * 40)

# All purchaser_id and owner_id exist in customers
clean_customer_ids = set(customers_df_clean['customer_id'].tolist())
orphan_check_purchaser = ticket_sales_df_clean[~ticket_sales_df_clean['purchaser_id'].isin(clean_customer_ids)]
orphan_check_owner = ticket_sales_df_clean[~ticket_sales_df_clean['owner_id'].isin(clean_customer_ids)]
print(f"\nTickets with invalid purchaser_id: {len(orphan_check_purchaser)}")
print(f"Tickets with invalid owner_id: {len(orphan_check_owner)}")
dq_results.append(("Consistency", "Tickets - All purchaser_id in customers", len(orphan_check_purchaser) == 0))
dq_results.append(("Consistency", "Tickets - All owner_id in customers", len(orphan_check_owner) == 0))

# --- DQ SUMMARY ---
# summary
print("\nResults")
print("-" * 40)
for category, check, passed in dq_results:
    status = "✓ PASS" if passed else "✗ FAIL"
    print(f"{status} | {category}: {check}")

1. Completeness
----------------------------------------

Tickets - Missing values: 0

Customers - Missing values by column:
country                         14
district                        28
ward                            28
distance_to_bramall_lane_km     28
age                            109
dtype: int64

2. Uniqueness
----------------------------------------

Tickets - Duplicate transaction_id: 0
Customers - Duplicate customer_id: 0

3. Validity
----------------------------------------

Tickets with transaction_date > event_date: 223


transaction_id  \
17   7387664d-4023-4424-aba5-23ec61cbf68a   
44   6e71ba8f-bd0a-4ddb-9019-2432106c178f   
107  35df980c-37ac-4096-8cd0-c85c967c2652   
135  bed6f217-37e2-434b-b189-fe73f0a26dc9   
146  a60dd63f-84e9-4272-a460-ef3887b29fbb   

                             purchaser_id  \
17   c1bce03b-7867-47b2-be87-58e4b380345e   
44   c1bce03b-7867-47b2-be87-58e4b380345e   
107  c1bce03b-7867-47b2-be87-58e4b380345e   
135  c1bce03b-7867-47b2-be87-58e4b380345e   
146  c1bce03b-7867-47b2-be87-58e4b380345e   

                                 owner_id transaction_date  \
17   c1bce03b-7867-47b2-be87-58e4b380345e       2024-08-21   
44   c1bce03b-7867-47b2-be87-58e4b380345e       2024-08-21   
107  c1bce03b-7867-47b2-be87-58e4b380345e       2024-08-21   
135  c1bce03b-7867-47b2-be87-58e4b380345e       2024-08-21   
146  c1bce03b-7867-47b2-be87-58e4b380345e       2024-08-21   

                      client_type  \
17   Match by Match Ticket Holder   
44   Match by Match Ticket Holder   
107  Match by Match Ticket Holder   
135  Match by Match Ticket Holder   
146  Match by Match Ticket Holder   

                              product_description event_date             area  \
17   2023/24 Sufc V Brighton & Hove Albion - Fac4 2024-01-27     The Pavilion   
44   2023/24 Sufc V Brighton & Hove Albion - Fac4 2024-01-27  1889 Restaurant   
107  2023/24 Sufc V Brighton & Hove Albion - Fac4 2024-01-27  1889 Restaurant   
135  2023/24 Sufc V Brighton & Hove Albion - Fac4 2024-01-27     The Pavilion   
146  2023/24 Sufc V Brighton & Hove Albion - Fac4 2024-01-27     The Pavilion   

    row_number seat_number  
17           Y         113  
44           L          11  
107          C           6  
135         AA         119  
146         AA         123


Customers with invalid age (outside 0-120): 0
Customers with negative distance: 0

4. Accuracy
----------------------------------------

Tickets with future transaction_date: 0
Tickets with future event_date: 5920

Transaction date range: 2024-07-17 00:00:00 to 2026-02-04 00:00:00
Event date range: 2024-01-27 00:00:00 to 2026-04-25 00:00:00

5. Consistency
----------------------------------------

Tickets with invalid purchaser_id: 0
Tickets with invalid owner_id: 0

Results
----------------------------------------
✓ PASS | Completeness: Tickets - No missing values
✓ PASS | Uniqueness: Tickets - No duplicate transaction_id
✓ PASS | Uniqueness: Customers - No duplicate customer_id
✗ FAIL | Validity: Tickets - transaction_date <= event_date
✓ PASS | Validity: Customers - Age in valid range (0-120)
✓ PASS | Validity: Customers - Distance is positive
✓ PASS | Accuracy: Tickets - No future transaction dates
✓ PASS | Consistency: Tickets - All purchaser_id in customers
✓ PASS | Consistency:

In [19]:
# Quick look at categorical distributions
print("Categorical distributions")
print("-" * 40)

# Tickets categorical columns
print("\n--- Tickets ---")
print(f"\nclient_type:\n{ticket_sales_df_clean['client_type'].value_counts()}")
print(f"\narea:\n{ticket_sales_df_clean['area'].value_counts()}")

# Customers categorical columns
print("\n--- Customers ---")
print(f"\ncustomer_type:\n{customers_df_clean['customer_type'].value_counts()}")
print(f"\ngender:\n{customers_df_clean['gender'].value_counts()}")
print(f"\ncountry:\n{customers_df_clean['country'].value_counts()}")
print(f"\nexclude_from_marketing:\n{customers_df_clean['exclude_from_marketing'].value_counts()}")

# Numerical summary for customers
print("\nNumerical summary — Customers")
print(customers_df_clean[['distance_to_bramall_lane_km', 'age']].describe())

Categorical distributions
----------------------------------------

--- Tickets ---

client_type:
client_type
Season Ticket Holder            32816
Match by Match Ticket Holder    10473
Name: count, dtype: int64

area:
area
The Pavilion         23657
Tony Currie Suite    10809
1889 Restaurant       8823
Name: count, dtype: int64

--- Customers ---

customer_type:
customer_type
Customer    202
Account      75
Name: count, dtype: int64

gender:
gender
Male           145
Unspecified    113
Female          19
Name: count, dtype: int64

country:
country
England           248
United Kingdom     10
Australia           2
Israel              1
Italy               1
Wales               1
Name: count, dtype: int64

exclude_from_marketing:
exclude_from_marketing
False    168
True     109
Name: count, dtype: int64

Numerical summary — Customers
       distance_to_bramall_lane_km         age
count                   249.000000  168.000000
mean                     26.808072   57.196429
std            

## Cleaning summary

What was done:
- Converted `transaction_date` and `event_date` to datetime
- Confirmed no duplicate customer IDs or transaction IDs
- Added `is_england_customer` flag for the 28 non-England customers (rather than dropping them)
- Kept all records with missing age — Accounts don't need age, and only 21 individual Customers have it missing (1.86% of tickets)
- Validated: no orphan tickets, all ages 0–120, all distances positive, transaction dates don't exceed event dates

Both datasets kept at 100% — 277 customers, 43,289 tickets. Nothing was dropped.

### On non-England customers
Excluding them would bias toward local fans. Instead they're flagged and can be filtered for distance-based work. District/ward/distance fields are England-specific, so those analyses just need `is_england_customer == True`.

### DQ monitoring suggestion
For ongoing data loads, worth implementing automated checks — things like uniqueness on IDs, null thresholds on age (for Customers specifically), referential integrity between tickets and customers. Something like Great Expectations would work well here. The key metrics to track: % missing age for Customers (currently 13.4%), % unmatched ticket IDs (currently 0%), % marketing opt-outs (for reach planning).